core/src/impl/Kokkos_Core.cpp

```diff
@@ -533,49 +533,75 @@ void initialize(int& narg, char* arg[])
+            Impl::throw_runtime_exception("Error: KOKKOS_DEVICE_ID out of range of representable values by an integer. Raised by Kokkos::initialize(int narg, char* argc[]).");
         if ((device != -1)&&(env_device!=device))
             Impl::throw_runtime_exception("Error: expecting a match between --kokkos-device and KOKKOS_DEVICE_ID if both are set'. Raised by Kokkos::initialize(int narg, char* argc[]).");
         else
             device = env_device;
     }
     auto env_ndevices_str = std::getenv("KOKKOS_NUM_DEVICES");
-    if (env_ndevices_str) {
-        auto env_ndevices = std::atoi(env_ndevices_str);
+    if (env_ndevices_str!=nullptr) {
+        errno = 0;
+        auto env_ndevices = std::strtol(env_ndevices_str,&endptr,10);
+        if (endptr== env_ndevices_str) 
+            Impl::throw_runtime_exception("Error: cannot convert KOKKOS_NUM_DEVICES to an integer. Raised by Kokkos::initialize(int narg, char* argc[]).");
+        if (errno == ERANGE && (env_ndevices == LONG_MAX || env_ndevices == LONG_MIN))
+            Impl::throw_runtime_exception("Error: KOKKOS_NUM_DEVICES out of range of representable values by an integer. Raised by Kokkos::initialize(int narg, char* argc[]).");
         if ((ndevices != -1)&&(env_ndevices!=ndevices))
             Impl::throw_runtime_exception("Error: expecting a match between --kokkos-ndevices and KOKKOS_NUM_DEVICES if both are set'. Raised by Kokkos::initialize(int narg, char* argc[]).");
         else
             ndevices = env_ndevices;
         //Skip device
         auto env_skip_device_str = std::getenv("KOKKOS_SKIP_DEVICE");
-        if (env_skip_device_str) {
-            auto env_skip_device = std::atoi(env_skip_device_str);
+        if (env_skip_device_str!=nullptr) {
+            errno = 0;
+            auto env_skip_device = std::strtol(env_skip_device_str,&endptr,10);
+            if (endptr== env_skip_device_str) 
+                Impl::throw_runtime_exception("Error: cannot convert KOKKOS_SKIP_DEVICE to an integer. Raised by Kokkos::initialize(int narg, char* argc[]).");
+            if (errno == ERANGE && (env_skip_device == LONG_MAX || env_skip_device == LONG_MIN))
+                Impl::throw_runtime_exception("Error: KOKKOS_SKIP_DEVICE out of range of representable values by an integer. Raised by Kokkos::initialize(int narg, char* argc[]).");
             if ((skip_device != 9999)&&(env_skip_device!=skip_device))
                 Impl::throw_runtime_exception("Error: expecting a match between --kokkos-ndevices and KOKKOS_SKIP_DEVICE if both are set'. Raised by Kokkos::initialize(int narg, char* argc[]).");
             else
                 skip_device = env_skip_device;
         }
     }
     const char * env_disablewarnings_str = std::getenv("KOKKOS_DISABLE_WARNINGS");
-    if (env_disablewarnings_str) {
+    if (env_disablewarnings_str!=nullptr) {
         if (strcmp(env_disablewarnings_str,"TRUE") == 0)
```

Comment:
Would be nice to do case-insensitive comparison, and to accept 0 or 1, ON or OFF, as well as TRUE or FALSE.  Right now, if you give this "true", you'll get `false`.